### PR TITLE
add metrics for harbor

### PR DESCRIPTION
**Pull request by ninjadq**

### CODE REVIEW

**Comment by MartinForReal** [comment]
related: https://github.com/goharbor/community/pull/103. I think comments provides a lot of useful information. FYI

File: proposals/expose-metrics.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+* Implementing service discovery.
+
+## Rationale
+
+* Why not expose metrics to Promethues Harbor components directly
+
+  * All Harbor components are behind a nginx proxy, it needs to configured to expose metrics for each components
+  * Prometheus needs to configure targets for each endpoint for scraping metric, but all of them are belong to harbor
+  * Use multiple metric APIs for one product seems not a good practice
```
Comment (simonpasquier) [inline proposals/expose-metrics.md:75]: From a Prometheus point of view, it's recommended that Prometheus scrapes each component directly. Similarly the Prometheus server should live as close as possible to its monitored targets.

File: proposals/expose-metrics.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+* Expose the metrics in prometheus format
+  * Use prometheus go SDK to expose runtime metrics like memory used
+  * Use a go routing to run in the background to periodically update business related metrics
+  * Write a middleware for all api handler to get the request performance data
+  * The port of exposing metrics should different from the main service
+* The architecture about Harbor standalong environment looks like below.
+  * ![Image of Harbor Metrics](images/observability/harbor-prom.png)
+  * Enable Prometheus format metrics in Harbor components 
+    * Only enabling Core in stage 1
```
Comment (xaleeks) [inline proposals/expose-metrics.md:45]: what is the reason for this?

File: proposals/expose-metrics.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    * Create a new container for OpenTelemetry collector
+    * Add labels to harbor components to distinguish them
+  * A Open Telemetry collector included in the Harbor package
+    * To collecting metrics from Harbor components
+    * Exposing them as a whole to the outside.
+  * Add new rule in nginx to expose metrics
+  * A new exporter component included in Harbor package
+    * colletct business logic data from DB
+    * Expose them as premetheus metric
+
```
Comment (xaleeks) [inline proposals/expose-metrics.md:58]: Can you add what a user of Harbor will need to do to consume these metrics.  ie how does he setup Prometheus, how does he subscribe to endpoints and metrics selectively, how does he request for additional metrics to be captured? @ninjadq

File: proposals/expose-metrics.md (hunk 1 above)
Comment (ninjadq) [reply proposals/expose-metrics.md:85]: yes, there are some update recently. We use `nginx` as reverse proxy to redirect prometheus scrape request to each component now.

File: proposals/expose-metrics.md (hunk 2 above)
Comment (ninjadq) [reply proposals/expose-metrics.md]: More component will added in the future.

File: proposals/expose-metrics.md (hunk 3 above)
Comment (ninjadq) [reply proposals/expose-metrics.md:68]: yup, I think we can add these kind of thing to document after code is freese.

**Comment by ninjadq** [comment]
@MartinForReal, it is really appreciate your proposal, we've read your proposal before. But in this proposal, we are focus on enabling the metrics on harbor. And provide an overall scaffold for people to add metrics. We will add more metrics in next steps, and will considering the proposal you posted.

**Review by wy65701436 — APPROVED** [review]
lgtm, let's merge the implemented feature.

**Comment by perithompson** [comment]
Hi All, is this PR still being looked at it seems to have gone quiet?

**Comment by steven-zou** [comment]
Lazy consensus and feature are delivered into 2.2.

**Pull request merged**